insalan/tournament/views.py

```diff
@@ -309,8 +319,89 @@ class TeamDetails(generics.RetrieveUpdateDestroyAPIView):
 
     queryset = Team.objects.all().order_by("id")
     serializer_class = serializers.TeamSerializer
-    permission_classes = [permissions.IsAdminUser | ReadOnly]
+    permission_classes = [permissions.IsAdminUser | ReadOnlyPlusPatch]
+
+    def patch(self, request, *args, **kwargs):
+        """
+        Patch a team
+        """
+        user = request.user
+        data = request.data
+
+        if user is None or not user.is_authenticated:
+            raise PermissionDenied()
 
+        # get the team
+        team = Team.objects.get(id=kwargs["pk"])
+
+        # check if the user is registered in the team
+        player = Player.objects.filter(user=user, team=team)
+        manager = Manager.objects.filter(user=user, team=team)
+        if len(player) == 0 and len(manager) == 0:
+            return Response({
+                "team": _("Vous n'êtes pas inscrit dans cette équipe.")
+            }, status=status.HTTP_403_FORBIDDEN)
+
+        # check if the player is the team's captain or a manager
+        if len(manager) == 0 and team.get_players_id()[0] != player[0].id:
+            return Response({
+                "team": _("Vous n'avez pas la permission de modifier cette équipe.")
+            }, status=status.HTTP_403_FORBIDDEN)
+
+        # team name edit
+        if "name" in data and len(data["name"]) >= 3:
+            team.name = data["name"]
+
+        # team password edit
+        if "password" in data:
+            if len(data["password"]) >= 8:
+                team.password = make_password(data["password"])
+            else:
+                return Response({
+                    "password": _("Mot de passe invalide.")
+                }, status=status.HTTP_400_BAD_REQUEST)
+
+        # player edit
+        if "players" in data:
+            existing = set(team.get_players_id())
+            players = set(data["players"])
+            removed = existing - players
+            for uid in removed:
+                player = Player.objects.get(id=uid)
+                # if player hasn't paid, remove him from the team
+                if player.as_user().id != user.id and player.payment_status == PaymentStatus.NOT_PAID:
+                    UserMailer.send_kick_mail(player.as_user(), team.name, user.username)
+                    player.delete()
```

Comment:
Do we fail silently if trying to kick a player who paid? Is that a desired behavior? (the same apply for managers and substitutes)

Reply:
That was desired yes because if we have to check before editing the team it would be costly. But yes, it could be a good idea to fail with an error.